File: Huxley/683DFSImprimindoAsFolhas.py
def floodFill(graph, visited, leafs, root, level):
    if (visited[root]): return
    visited[root] = 1
    leafs[level] += [root]
    if (len(graph[root]) == 0): return
    for u in graph[root]:
        floodFill(graph, visited, leafs, u, level + 1)

while (True):
    try:
        lines = int(input())
    except:
        break
    graph, first = {}, -1
    for i in range(lines):
        u, v = input().split()
        if (first == -1): first = u
        if (u not in graph):
            graph[u] = []
        if (v not in graph):
            graph[v] = []
        graph[u] += [v]
    print("graph", graph)

    leafs = [[] for i in range(len(graph) + 1)]
    visited = {}
    for i in graph:
        visited[i] = 0
    for i in graph:
        if (not visited[i]):
            floodFill(graph, visited, leafs, i, 0)
            print(visited)

    allLeafs = []
    for i in range(len(leafs)):
        leafs[len(leafs) - i - 1].sort()
        allLeafs += leafs[len(leafs) - i - 1]
    print("allLeafs", allLeafs)
    # for i in leafs:
    #     if (len(i) == 0):
    #         break
    #     for j in i:
    #         print(j)

File: Huxley/683DFSimprimindoAsFolhas.cpp
#include <bits/stdc++.h>
using namespace std;
map<string, int> stringToInt;
map<int, string> intToString;

int dfs(vector<vector<int> > graph, int u, int prev, int depth)
{
  if (!graph[u].size()) return(depth);
  int ans = 0;
  for (auto v: graph[u])
    if (v != prev)
      ans = max(ans, dfs(graph, v, prev, depth + 1));
  return(ans);
}

int leafPerLevel(vector<vector<int> > graph, int u, int prev, set<int>& level, int visited[])
{
  int some = 0;
  for (auto v: graph[u])
    if (prev != v && !visited[u])
    {
      leafPerLevel(graph, v, prev, level, visited);
      some = 1;
    }
  if (!some)
  {
    level.insert(u);
    visited[u] = 1;
  }
  return(0);
}

int main()
{
  int lines;
  while (scanf("%d", &lines) != EOF)
  {
    vector<vector<int> > graph;
    char u[100], v[100];
    for (int i = 0; i < lines; i ++)
    {
      scanf("\n%s %s", u, v);
      if (!stringToInt.count(u))
      {
        intToString[stringToInt.size()] = u;
        stringToInt[u] = stringToInt.size();
      }
      if (!stringToInt.count(v))
      {
        intToString[stringToInt.size()] = v;
        stringToInt[v] = stringToInt.size();
      }
      while (graph.size() < stringToInt.size()) graph.push_back(vector<int>());
      graph[stringToInt[u]].push_back(stringToInt[v]);
      graph[stringToInt[u]].push_back(stringToInt[v]);
    }
    int depth = dfs(graph, 0, 0, 0);
    set<int> level[depth + 1]; int visited[(int) stringToInt.size()]; memset(visited, 0, sizeof(visited));
    for (int i = 0; i <= depth; i ++)
      leafPerLevel(graph, 0, 0, level[i], visited);
    set<string> ans[depth + 1];
    for (int i = 0; i <= depth; i ++)
      for (auto j: level[i])
        ans[i].insert(intToString[j]);
    for (int i = depth; i >= 0; i --)
      for (auto j: ans[i])
        printf("%s\n", j.c_str());
  }

  return(0);
}